File: 0039-combination-sum/0039-combination-sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        final = []
        def combo(value):
            for c in candidates:
                copy = list(value)
                copy.append(c)
                copy.sort()
                if sum(copy) < target:
                    combo(copy)
                elif sum(copy) == target and copy not in final:
                    final.append(copy)
        combo([]);
        return final

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    ArrayList<List<Integer>> answer = new ArrayList<>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<Integer> empty = new ArrayList<>();
        Arrays.sort(candidates);
        checker(candidates, target, empty, 0);
        return answer;
    }
    
    public void checker(int[] candidates, int target, List<Integer> current, int index){
        if(target == 0){
            if(!answer.contains(current)){
                answer.add(new ArrayList<>(current));
            }
        }
        else if(target > 0){
            for(int i = index; i< candidates.length; i++){
                if(candidates[i] <= target){
                current.add(candidates[i]);
                checker(candidates, target - candidates[i], current, i);
                current.remove(current.size()-1);
            }
            }
        }
    }
}